### ISSUE TITLE
perf(speculative): attribute the Laguna DFlash verify round's fixed and per-row cost on GB10

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1799]
## Problem / Background

On the Laguna NVFP4 DFlash pairing (`models/mlx/laguna-xs-2.1-nvfp4` target with `models/mlx/laguna-xs-2.1-dflash` drafter), a multi-row speculative verify round still costs a fixed floor plus a per-row term that neither PR #1795 fix removes, so speculative decoding has no net win at any block width on GB10. PR #1771 (issue #1351) is blocked on this issue by maintainer decision: the Laguna DFlash drafter is implemented, reviewed and CI-green at `dab18fdf` (every check SUCCESS or SKIPPED) behind a fail-closed exactness gate, and is held unmerged until the pairing shows a net win here. The hold is not free: the branch has already taken two merge commits to follow main (`d64c73af`, `08e09036`) and will take more the longer it waits.

Post-rebase sweep recorded on PR #1771 and in `TECHNICAL_REPORTS/1351-laguna-dflash-drafter-20260911.en.md:59-83` (branch-only). GB10 sm_121, MLX pin `81ba1c6a` (`src/lib/mlx-cpp/CMakeLists.txt:122`), CUDA release build, same binary with the feature off as the control arm, GPU held exclusively, n = 3 per width, 200-token raw code completions through `mlxcel generate`, decode tok/s as the CLI reports it.

| width | tok/s mean (min to max) | mean accepted | device ms/round | drafter host build ms/round | vs off |
|---|---|---|---|---|---|
| off | 31.18 (30.41 to 31.98) | | | | |
| 2 | 16.19 (16.12 to 16.25) | 0.84 | 80 | 30 | 0.52x |
| 4 | 27.05 (26.61 to 27.53) | 2.21 | 85 | 31 | 0.87x |
| 6 | 30.46 (30.05 to 31.26) | 2.90 | 92 | 33 | 0.98x |
| 8 | 31.26 (31.07 to 31.36) | 3.33 | 101 | 34 | 1.00x |
| 10 | 30.02 (29.68 to 30.19) | 3.42 | 109 | 35 | 0.96x |
| 16 (default) | 27.41 (26.90 to 27.90) | 3.55 | 127 | 35 | 0.88x |

Best width is 8 at 1.00x with its whole range inside the off arm's, so it is not a demonstrated win, and the checkpoint default of 16 is 0.88x. The PR text fits the device time as a fixed 77 ms plus about 3.1 ms per row; an ordinary least squares fit over all eleven measured widths gives 72 ms plus 3.55 ms per row. Either way the floor alone is more than two classic tokens: 80 ms at 2 rows against a 31 to 32 ms graph-replayed classic token is 2.5x a single-token step. Including the 34 ms of host-side drafter graph construction, perfect acceptance at width 8 would reach about (101 + 34) / 8 = 17 ms per token, capping near 1.8x to 1.9x; at the measured 3.33 accepted it lands at 1.00x.

## Why neither #1795 fix reaches this pairing

PR #1795 removed two causes on the Qwen 3.5 pairing: an unconditional bf16 to f16 drafter conversion that promoted the whole drafter to float32, and a row-independent 64-row gated-delta chunked scan. Both statements below were checked in the code, not inherited from the PR text. First, the Laguna drafter was already bf16 on this host before #1795: the loader carried its own `drafter_weights_prefer_f16()` with the same Ampere-and-later rule, removed in `aad08e1c` in favor of the shared policy, so `src/lib/mlxcel-core/src/drafter/laguna_dflash/drafter.rs:65` now calls `apply_drafter_load_dtype_policy`, which returns false on compute capability 12.1 unless `MLXCEL_CUDA_F16_NORMALIZE` is set (`src/lib/mlxcel-core/src/drafter/dflash/drafter.rs:187-191,241-270`). Second, Laguna has no gated-delta layers: it is mixed full and sliding softmax attention with optional per-head sinks (`src/models/laguna_layers.rs:110-199` on the PR branch), and `gated_delta` appears nowhere in the Laguna sources. The measured numbers duly did not move across the rebase.

## The mechanism is NOT established, and that is the point of this issue

The phrase used so far, "the classic step is graph-replayed while the multi-row verify runs eagerly and launch-bound", is an inference from the fixed-plus-per-row arithmetic fit. It is not a profiled result: no nsys kernel attribution has been run on the Laguna pairing at any width. Do not implement against it. The Qwen 3.5 investigation in #1782 disproved exactly that intuition on the pairing where it was actually measured: CUDA graph capture is on for every eval unless `MLX_USE_CUDA_GRAPHS=0` (MLX pin, `mlx/backend/cuda/device.cpp:17-18`, default true) and nothing in the capture path branches on row count; replay was healthy at block 2 (260 `cudaGraphInstantiate` against 71089 `cudaGraphExecUpdate`); turning capture off made block 8 6 percent faster; and the classic step turned out GPU-bound on the quantized projections (12.4 ms of about 15.5 ms of per-token kernel time in `docs/benchmark_results/dflash-verify-fixed-cost-gb10-2026-09-11.md`), not launch-bound. The real causes there were a dtype bug and an algorithmic one, and neither was visible in the arithmetic fit.

## Proposed Solution

Attribution first, implementation second, the shape that worked for #1782.

**Phase 0, attribution (required, and the only phase with a fixed deliverable).** Profile the Laguna verify round per block width with `nsys profile -t cuda,nvtx --cuda-graph-trace=node` (mandatory: without it graph-captured work lands on `event_signal_kernel`, rule 4 of `docs/benchmark_results/volta-sm70-baseline-2026-08-31.md:64-88`). Report kernel names, absolute kernel time and per-round launch counts at 2, 8 and 16 rows, and split them into the kernels whose counts and time stay flat across widths (the fixed floor) and those that scale with rows (the per-row term). Reconcile the profiled run against an unprofiled one and state the inflation ratio before drawing any kernel-level conclusion. Name the mechanism, with the profile as the evidence.

Candidates to check. Each needs evidence; none may be assumed.

- **Is the multi-row verify replayed or re-instantiated on this pairing?** Measure it as the `cudaGraphInstantiate` to `cudaGraphExecUpdate` ratio per arm, the same way #1782 did. MLX keys an LRU of instantiated graphs (`graph_cache_`, default capacity 400, `mlx/backend/cuda/device.cpp:210`) and reuses through `cudaGraphExecUpdate` (`device.cpp:527-539`), so a round whose shapes vary can miss the cache here even though it did not on Qwen 3.5. A partial accept changes the next round's row count, so this is a live possibility rather than a formality.
- **The NVFP4 dispatch path.** `supports_fp_qmv` refuses when `vec_batch > 8`, and also refuses affine mode and compute capability 9 or below (MLX at the pin, `mlx/backend/cuda/quantized/qmm/qmm.cu:127-157`), while `quantized.cpp` selects `qmv` or `fp_qmv` when `M * B < 8` and `qmm_sm80` otherwise (`src/lib/mlx-cpp/patches/mlx/backend/cuda/quantized/quantized.cpp:282-292` dense, `:525-534` gather). So the family switch at 8 rows is real on this pairing, but the 5-to-7 accumulator cliff seen on the affine path does not exist here: `fp_qmv.cu` instantiates only `fp_qmv_single` and `fp_qmv_batched`, its `rows_per_block = 8` (`fp_qmv.cu:17`) tiles output rows (`blocks_y = (N + rows_per_block - 1) / rows_per_block`), and the only per-launch template variation is `n_per_thread` in {1, 2, 4} chosen by pointer alignment, not by input rows. The measured curve agrees: 5, 6 and 7 sit on a smooth line. Note also that Laguna is MoE and its experts go through `SwitchGLU` and `gather_qmm` (`src/models/laguna_layers.rs:23,451-469`), where there is no `fp_qmv` arm at all (`quantized.cpp:495-534`) and the fused decode-MoE kernel does not apply to a non-affine expert set (`src/models/switch_layers.rs:79-81`), so the expert matmuls are a separate dispatch question from the attention projections and the profile must separate them.
- **The Laguna attention shape itself**, including the multi-row mask. A verify with `l > 1` materializes an additive array mask per layer, a causal band on full layers and a window-clipped band on sliding layers, while a single-token decode passes no mask at all (`src/models/laguna_layers.rs:165-199`). On Qwen 3.5 the equivalent materialized-mask term measured under 2 ms per round at 16 rows and was ruled out; here it is unmeasured.
- **Drafter host-side graph construction**, 30 to 35 ms per round at every width, roughly one classic token, and a serial link separate from device time (`diagnostics.draft_time_ms`, built at `src/lib/mlxcel-core/src/drafter/dflash/round_loop.rs:754-767` and logged as `draft_ms` by `src/server/batch/dflash_target.rs:706-718`).

**Phase 1, change, only if phase 0 supports one.** A round is a serial chain of host draft build, host verify build and device sync (`round_loop.rs:751-818`), which is why the two #1795 fixes were superlinear together: each removed a different link and either alone left the others dominating. Measure any proposed change against the whole round and against end-to-end throughput, never against the link it targets.

**Phase 2, re-measurement.** Same method as the sweep above and as #1782: same binary on every arm, feature off as the control, GPU held exclusively, host load stated, n >= 3 per arm with min and max reported, and a width range wide enough to find a crossover if one exists (2 through 16 at least). Commit the record under `docs/benchmark_results/` and the attribution under `TECHNICAL_REPORTS/`.

## Scope

**In scope:** the per-round cost of the Laguna DFlash verify on CUDA, its attribution, and any change the attribution justifies in `src/lib/mlxcel-core/src/drafter/dflash/round_loop.rs`, `src/models/laguna_speculative.rs`, `src/models/laguna_layers.rs` or the MLX patches under `src/lib/mlx-cpp/patches/`; the committed benchmark record and technical report.

**Out of scope:**

- The default draft block width and any hardware-gated width policy: that is #1797, which already names this pairing as unmeasurable until #1771 lands, and already puts the `qmm_sm80` term at 8 rows and above outside its own scope.
- The second unmet criterion on PR #1771, the exactness probe, which belongs in its own issue and must not be folded into a performance negative. Recorded here so it is not lost: the probe declines this host by default at width 16 (107246 of 200704 logit bytes differ at the first verify position) and at width 8 (107466), the `qmv_wide` retry does not change it, and with `MLXCEL_MTP_ALLOW_INEXACT=1` the divergences are bf16 tie flips, 1 of 6 prompts byte-identical and 1 to 2 tie positions per 128 tokens on the other 5, every one at a chain-arm top-2 margin of 0.0 or 0.125. It is kept separate because its mechanism (kernel rounding at logit ties) and its fix surface (the gate in `src/models/laguna_speculative.rs:98-135` and the kernel numerics) share nothing with device-time attribution, and because folding it in would let a measured performance negative close an open correctness question. One concrete lead for that issue: both recorded declines are at widths where `M * B >= 8` sends the projections to `qmm_sm80`, while widths 2 to 7 stay on `fp_qmv`, whose per-row arithmetic is row-independent by construction; the gate probes before it consults the override (`src/models/speculative_exactness.rs:295-312`), so per-width verdicts at 2 to 7 exist in the sweep logs or can be re-run cheaply.

## Implementation Notes

- **Reuse**: the per-round split is already instrumented. `DFlashDiagnostics` carries `draft_time_ms`, `verify_time_ms` and `target_argmax_time_ms` (`round_loop.rs:530-538`) and the server logs them as `draft_ms`, `verify_ms` and `target_argmax_sync_ms` on the `DFlash diagnostics` line (`dflash_target.rs:706-718`). Do not add a parallel timing path. `scripts/bench_block_width.sh` already interleaves widths so drift spreads across the table, but it hardcodes `--draft-kind mtp` (lines 86 and 112) and drives `mlxcel generate`; extend it rather than writing a new harness, and say which harness produced each number.
- **Constraints**: the fix must not change greedy output on any pairing, must keep a same-binary kill switch for the A/B as #1795 did, and must not regress the classic arm or the Qwen 3.5 speculative arm (1.24x at width 2 and 1.32x at 4 after #1795). Anything touching the MLX patches under `src/lib/mlx-cpp/patches/` reaches every quantized decode on CUDA, so state the blast radius.
- **Measurement hazards**: `MLXCEL_TRACE_ARCH` prints the quantized path taken on the process's first call only (`quantized.cpp:58-74`), which is prefill, so it cannot separate verify widths; use nsys kernel names instead. GB10 single-run decode is bimodal by up to 25 percent even at pinned clocks (#755), hence n >= 3 and reported spreads. The #1771 sweep ran with the GPU exclusive but the host not idle (another session's `cargo build` at load 2.6 to 3.9), which is a real hazard for a chain whose drafter graph build is host-side, so re-measurement must run on an idle host and say so. nsys inflated the #1782 runs by 15 to 46 percent, so compare proportions and launch counts across arms, not absolute times.
- **Edge cases**: a partial accept makes the next round's row count vary, which is what could churn graph keys; rollback replays accepted rows and can add launches the width alone does not predict; the sliding caches run with speculative slack (`src/models/laguna_speculative.rs:64-82`), so the key count a rotating layer exposes changes as the window fills.

## Acceptance Criteria

- [x] Phase 0 names the mechanism of the fixed floor and of the per-row term, each backed by a committed nsys table with kernel names and per-round launch counts at 2, 8 and 16 rows, stating which kernels are flat across widths and which scale.
- [x] The graph replay question is answered with a number: the `cudaGraphInstantiate` to `cudaGraphExecUpdate` ratio on the Laguna verify arm at at least two widths.
- [x] Every candidate above is either confirmed with evidence or ruled out with a measured bound, including the ones that turn out not to matter.
- [x] Any code change ships with a same-binary kill switch, and its A/B is reported against the whole round and against end-to-end throughput, not against the link it targets.
- [x] Re-measurement across widths 2 to 16 on an idle host with the GPU held exclusively, n >= 3, min and max reported, same binary and feature-off control, with the classic arm re-measured in the same session.
- [x] Greedy output is unchanged on the Qwen 3.5 pairing (still byte-identical to classic at every width after #1795) and that pairing's speculative win is not regressed.
- [x] A benchmark record under `docs/benchmark_results/` and a technical report under `TECHNICAL_REPORTS/` (`.en.md` and `.ko.md`), following the #1782 pair.
- [x] **A measured negative closes this issue successfully.** (Not needed: the pairing reaches a net win, 1.33x at block 4. Kept ticked as the clause is satisfied by the positive result.)  If the evidence says the cost is irreducible on this pairing and this host, record that with the profile, say what would have to change (hardware, kernel, or drafter shape) for the conclusion to flip, and hand the merge decision on PR #1771 back to the maintainer with that evidence. Shipping a change that does not move the measurement is not an acceptable outcome.

## Verification

```
# Attribution (per width; repeat at 2, 8, 16)
nsys profile -t cuda,nvtx --cuda-graph-trace=node -o laguna_w8 \
  mlxcel generate -m models/mlx/laguna-xs-2.1-nvfp4 \
    --draft-model models/mlx/laguna-xs-2.1-dflash --draft-kind dflash \
    --draft-block-size 8 -n 200 --temp 0

# Per-round split from the same runs
RUST_LOG=info mlxcel generate ... 2>&1 | grep "DFlash diagnostics"

# Graph replay ratio
nsys stats --report cuda_api_sum laguna_w8.nsys-rep | grep -E "cudaGraphInstantiate|cudaGraphExecUpdate"

# Gate (this host)
cargo fmt --all -- --check
cargo clippy --profile test-fast --features cuda --lib --bins --tests -- -D warnings
make verify-test-cuda   # cargo test --workspace --profile test-fast --features cuda -- --test-threads=1
```

A pass is: the profile tables committed, the fixed and per-row terms each attributed to named kernels, and either a re-measured throughput table showing a net win at some width or a recorded negative with the evidence behind it.

## Technical Considerations

Provenance: issue #1351 and PR #1771 (the drafter, blocked on this), issue #1782 and PR #1795 (the Qwen 3.5 attribution and fix, merged as `e391ae9c`), issue #1797 (the width policy follow-up). Prior art to read first: `docs/benchmark_results/dflash-verify-fixed-cost-gb10-2026-09-11.md` and `TECHNICAL_REPORTS/1782-dflash-verify-fixed-cost-cuda-20260911.en.md`. The Laguna-side numbers quoted above live on the PR branch in `TECHNICAL_REPORTS/1351-laguna-dflash-drafter-20260911.en.md` and in `docs/supported-models.md:714`, which already documents the pairing as measured and not winning on GB10.

## Resolution log (2026-09-11)

Attribution on an idle host, tree `2deae307`: the fixed floor is host time in `ScaledDotProductAttention::eval_gpu` (67 to 76 ms per round against 2.8 ms per classic token, nsys NVTX ranges differenced per round): a multi-row masked SDPA with head_dim 128 is cuDNN-eligible, MLX caches the cuDNN plan by exact `k_len` and mask shape, and a verify round's key length is new every round, so three shape classes (target full, target sliding with sinks, drafter) rebuild a plan each round at about 22 ms; the same LRU's lifetime miss counter aborted every block-2 run past about 170 rounds. The per-row term is the routed experts reading each selected expert per row (`qmm_sm80`, 2.5 ms per row). Ruled out with bounds: byte and op budgets, capture, re-instantiation (under 1 instantiate per 1000 exec updates), the mask (under 1 ms of GPU time per round), and the NVFP4 family switch on the projections, which does not exist here because this checkpoint's attention projections are bf16 and `gather_qmm` sees `M = 1, B = 8n` so the experts take `qmm_sm80` at every width. Fix: a patched `supports_sdpa_cudnn` that keeps 2 to 32 query rows over a longer key sequence off cuDNN (`MLXCEL_SDPA_FALLBACK_MAX_QUERIES`, 0 restores upstream), plus `MLX_CUDA_SDPA_CACHE_SIZE=2000` by default on CUDA. After, same binary kill-switch A/B, n = 3: block 4 1.33x, block 6 1.32x, block 2 1.13x, block 8 1.12x, crossover between 10 and 12, checkpoint default 16 at 0.83x; Qwen 3.5 unchanged (1.26x, 1.35x) and byte-identical. Record: `docs/benchmark_results/laguna-dflash-verify-cost-gb10-2026-09-11.md`; report `TECHNICAL_REPORTS/1799-laguna-dflash-verify-cost-cuda-20260911.{en,ko}.md`. Also recorded for #1798: GB10's default graph budgets make capture a net loss on this MoE classic arm (graphs off +12%, both budgets raised +16%), opposite sign to Qwen 3.5.
